src/include/txn_inline.h

```diff
@@ -310,26 +311,64 @@ __wt_txn_op_delete_commit_apply_timestamps(WT_SESSION_IMPL *session, WT_REF *ref
      */
     if (previous_state == WT_REF_DELETED) {
         if (ref->ft_info.del->timestamp == WT_TS_NONE) {
-            ref->ft_info.del->timestamp = txn->commit_timestamp;
-            ref->ft_info.del->durable_timestamp = txn->durable_timestamp;
+            if (in_commit_phase) {
+                /*
+                 * Updates without a timestamp can be retroactively timestamped on transaction
+                 * commit, but if this happens for a partially timestamped transaction then using
+                 * commit_timestamp will lead to out-of-order timestamps. Using
+                 * first_commit_timestamp will guarantee a correct timestamp ordering.
+                 */
+                ref->ft_info.del->timestamp = txn->first_commit_timestamp;
+                ref->ft_info.del->durable_timestamp = txn->first_commit_timestamp;
+            } else {
+                ref->ft_info.del->timestamp = txn->commit_timestamp;
+                ref->ft_info.del->durable_timestamp = txn->durable_timestamp;
+            }
         }
     } else if ((updp = ref->ft_info.update) != NULL)
-        for (; *updp != NULL; ++updp) {
-            (*updp)->start_ts = txn->commit_timestamp;
-            (*updp)->durable_ts = txn->durable_timestamp;
-        }
+        for (; *updp != NULL; ++updp)
+            if ((*updp)->start_ts == WT_TS_NONE)
```

Comment:
This `== WT_TS_NONE` check is new, but matches the behaviour of all other locations where we can set `first_commit_timestamp`

Reply:
That is interesting as I think it implies that previously a truncates visibility could change on commit.